### ISSUE TITLE
Xamarin 0.8.5+ System.NotImplementedException: The method or operation is not implemented.   at System.Net.Sockets.SocketAsyncEventArgs.FinishOperationAsyncFailur

### ISSUE THREAD

**Issue by mattheworiordan** [ably/ably-dotnet#161]
Customer reported:

> I’m doing some testing and I think I found a bug in the newest version 0.8.5 and later for Xamarin
 
> I received it in my app and I was also able to reproduce the error in the Xamarin.Forms sample app (MonkeyChat)
 
> It is working in the older version: 0.8.4.8, but not 0.8.5 or latest beta version
 
> The error occurs when you have made a connection and attached channel then turn of Wifi so you don't have a internet connection.
Sometimes the error comes after a while (30 sec), or the error comes when you turn on the Wifi again
 
> The application stops with this error message:

```
2017-11-01 13:41:22.088 MonkeyChatiOS[37066:3517097]
Unhandled Exception:
System.NotImplementedException: The method or operation is not implemented.
  at System.Net.Sockets.SocketAsyncEventArgs.FinishOperationAsyncFailure (System.Exception exception, System.Int32 bytesTransferred, System.Net.Sockets.SocketFlags flags) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.2.0.11/src/mono/mcs/class/System/System.Net.Sockets/SocketAsyncEventArgs.cs:257
  at System.Net.Sockets.MultipleConnectAsync.AsyncFail (System.Exception e) [0x0001a] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.2.0.11/src/mono/mcs/class/referencesource/System/net/System/Net/Sockets/_MultipleConnectAsync.cs:305
  at System.Net.Sockets.MultipleConnectAsync.Fail (System.Boolean sync, System.Exception e) [0x0000c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.2.0.11/src/mono/mcs/class/referencesource/System/net/System/Net/Sockets/_MultipleConnectAsync.cs:273
  at System.Net.Sockets.MultipleConnectAsync.DoDnsCallback (System.IAsyncResult result, System.Boolean sync) [0x000b4] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.2.0.11/src/mono/mcs/class/referencesource/System/net/System/Net/Sockets/_MultipleConnectAsync.cs:133
  at System.Net.Sockets.MultipleConnectAsync.DnsCallback (System.IAsyncResult result) [0x00008] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.2.0.11/src/mono/mcs/class/referencesource/System/net/System/Net/Sockets/_MultipleConnectAsync.cs:78
  at (wrapper managed-to-native) System.Runtime.Remoting.Messaging.AsyncResult:Invoke (System.Runtime.Remoting.Messaging.AsyncResult)
  at System.Runtime.Remoting.Messaging.AsyncResult.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.2.0.11/src/mono/mcs/class/corlib/System.Runtime.Remoting.Messaging/AsyncResult.cs:210
  at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00074] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.2.0.11/src/mono/mcs/class/referencesource/mscorlib/system/threading/threadpool.cs:856
  at ObjCRuntime.Runtime.ThreadPoolDispatcher (System.Func`1[TResult] callback) [0x00006] in /Users/builder/data/lanes/5481/2f8bbec0/source/xamarin-macios/src/ObjCRuntime/Runtime.cs:244
  at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () [0x00009] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.2.0.11/src/mono/mcs/class/referencesource/mscorlib/system/threading/threadpool.cs:1208
 
  2017-11-01 13:41:22.089 MonkeyChatiOS[37066:3517097] Unhandled managed exception:
The method or operation is not implemented. (System.NotImplementedException)
  at System.Net.Sockets.SocketAsyncEventArgs.FinishOperationAsyncFailure (System.Exception exception, System.Int32 bytesTransferred, System.Net.Sockets.SocketFlags flags) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.2.0.11/src/mono/mcs/class/System/System.Net.Sockets/SocketAsyncEventArgs.cs:257
  at System.Net.Sockets.MultipleConnectAsync.AsyncFail (System.Exception e) [0x0001a] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.2.0.11/src/mono/mcs/class/referencesource/System/net/System/Net/Sockets/_MultipleConnectAsync.cs:305
  at System.Net.Sockets.MultipleConnectAsync.Fail (System.Boolean sync, System.Exception e) [0x0000c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.2.0.11/src/mono/mcs/class/referencesource/System/net/System/Net/Sockets/_MultipleConnectAsync.cs:273
  at System.Net.Sockets.MultipleConnectAsync.DoDnsCallback (System.IAsyncResult result, System.Boolean sync) [0x000b4] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.2.0.11/src/mono/mcs/class/referencesource/System/net/System/Net/Sockets/_MultipleConnectAsync.cs:133
  at System.Net.Sockets.MultipleConnectAsync.DnsCallback (System.IAsyncResult result) [0x00008] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.2.0.11/src/mono/mcs/class/referencesource/System/net/System/Net/Sockets/_MultipleConnectAsync.cs:78
  at (wrapper managed-to-native) System.Runtime.Remoting.Messaging.AsyncResult:Invoke (System.Runtime.Remoting.Messaging.AsyncResult)
  at System.Runtime.Remoting.Messaging.AsyncResult.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.2.0.11/src/mono/mcs/class/corlib/System.Runtime.Remoting.Messaging/AsyncResult.cs:210
  at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00074] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.2.0.11/src/mono/mcs/class/referencesource/mscorlib/system/threading/threadpool.cs:856
  at ObjCRuntime.Runtime.ThreadPoolDispatcher (System.Func`1[TResult] callback) [0x00006] in /Users/builder/data/lanes/5481/2f8bbec0/source/xamarin-macios/src/ObjCRuntime/Runtime.cs:244
  at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () [0x00009] in /Library/Frameworks/Xamarin.iOS.framework/Versions/11.2.0.11/src/mono/mcs/class/referencesource/mscorlib/system/threading/threadpool.cs:1208
```

**Comment by withakay** [comment]
I can confirm that this bug is repeatable and is a regression from 0.8.4.8.

**Comment by withakay** [comment]
I believe this is an upstream problem with websocket4net. See this [issue](https://github.com/kerryjiang/WebSocket4Net/issues/108).

Essentially 
System.Net.Sockets.SocketAsyncEventArgs.FinishOperationAsyncFailure() is being called but it is [not implemented in mono](https://github.com/mono/mono/blob/master/mcs/class/System/System.Net.Sockets/SocketAsyncEventArgs.cs).

The websocket4net author claims to have a fix but at the time of writing a new package has not been released.

**Comment by withakay** [comment]
If issue #151 is resolved the I think we could move ahead with dropping websocket4net as a dependency for Xamarin targets and use System.Net.WebSockets everywhere.

If people have a specific reason to use websockets4net we could provide a separate package that provides an implementation of ITransportFactory the can be plugged in via ClientOptions.

Does anyone have any thoughts on that?

**Comment by mattheworiordan** [comment]
> If people have a specific reason to use websockets4net we could provide a separate package that provides an implementation of ITransportFactory the can be plugged in via ClientOptions.

I can't see any reason why someone would need that.  As far as I was aware, we only supported both because of issues in the MS version, so if that is resolved, we can remove this dependency and it's a win for everyone.

**Comment by withakay** [comment]
> I can't see any reason why someone would need that.

There are scenarios, but they can probably be considered edge cases. For example, if a customer were stuck for some reason on an old version of Xamarin or mono? 

But, I am not suggesting this is something we would do in anticipation but rather something we could do as a response, if such a need materialises.

**Comment by mattheworiordan** [comment]
Yes, that makes sense.  If it’s an issue for customers, we can offer them a
way to inject the lib themselves right?

**Comment by withakay** [comment]
@mattheworiordan yes, exactly, we could either provide cut and paste code examples or make a separate nuget package that contains the functionality. The later could be made very simple for the end user, simply add the package and add one line of code to assign the TransportFactory via the ClientOptions.

**Comment by mattheworiordan** [comment]
Ok, well let's not publicise the options for now.  If customers hit issues, we can consider one of the two ideas you have.  Thanks!

**Comment by apertsovsky** [comment]
Any updates on issue?

**Comment by withakay** [comment]
@apertsovsky I am working on a new release that will no longer depend on websocket4net, which is the source of this issue.

**Comment by apertsovsky** [comment]
Isn't it a mono issue after all? Looks like they just copied MultipleConnectAsync from .net sources and just left methods that it uses unimplemented to be able to compile their project.

**Comment by withakay** [comment]
@apertsovsky well, I suppose that depends on how you look at it ;)
It is a mono issue causing that is causing an issue with websockets4net. In the mean time there have been updates to mono that mean that System.Net.WebSockets now works. We currently use System.Net.WebSockets in the .Net 4.6.1 and .Net Standard builds and would like that to be the same across all supported platforms. In general, we want to avoid having platform specific dependencies where possible.

**Comment by withakay** [comment]
This is now fixed on master, a 0.8.6-beta4 release is coming soon.
